README.md
# benchmark

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-09-19
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-09-19
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/benchmark/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/benchmark/compare/0.1.0...0.1.1


test/benchmark/core_test.clj
(ns benchmark.core-test
  (:require [clojure.test :refer :all]
            [benchmark.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/benchmark/sync.clj
 (ns benchmark.sync
  (:refer-clojure :exclude [sync])
  (:import
   ;;[com.mongodb.client MongoClients]
   ;;[com.mongodb.client MongoClient]
   [com.mongodb.client MongoCollection]
   [com.mongodb.client MongoDatabase]
   [com.mongodb MongoClientOptions MongoClient ServerAddress Block]
   [org.bson Document]
   ;;[com.mongodb Block ConnectionString MongoClientSettings ServerAddress]
   ;;[com.mongodb.connection ClusterSettings ConnectionPoolSettings]
   [java.util.concurrent CyclicBarrier CountDownLatch]
   ))


#_(def mongo-client (MongoClients/create
                   (-> (MongoClientSettings/builder)
                       (.applyToClusterSettings
                        (reify
                          Block
                          (apply [this x]
                            (-> x
                                (.hosts [(ServerAddress. "192.168.33.10" 27017)])))))
                       (.applyToConnectionPoolSettings
                        (reify
                          Block
                          (apply [this x]
                            (-> x

                                (.maxSize 100)
                                (.minSize 50)
                                (.maxWaitQueueSize 10000)))))
                       (.build))))

(def mongo-client (MongoClient. [(ServerAddress. "192.168.33.10" 27017)]))
(def all-run-times 10000)
(def threads 20)
(def run-times (/ all-run-times threads))
(def test-db "test-db")
(def test-class "test-class")

(def default-blockcb (reify
                       Block
                       (apply [this s])))

(defn drop-all
  []
  (let [db-collection (-> (.getDatabase mongo-client test-db)
                          (.getCollection test-class))]
    (println "start drop data.")
    (.drop db-collection)))

(defn sync-query
  []
  (let [db-client (.getDatabase mongo-client test-db)
        query (Document. {"name" "_User"})
        start (System/currentTimeMillis)]
    (dotimes [_ run-times]
      (-> db-client
          (.getCollection "_Class")
          (.find query)
          (.forEach default-blockcb)))
    (println  (format "sync query %s times user %s ms" run-times (- (System/currentTimeMillis) start)))))

(defn sync-insert
  [id cdl]
  (let [db-client (.getDatabase mongo-client test-db)
        start (System/currentTimeMillis)]
    (dotimes [i run-times]
      (-> db-client
          (.getCollection test-class)
          (.insertOne (Document. {"name" (str (rand-int 10000))
                                  "_id" (str (+ i id))}))))
    (.countDown cdl)
    (println  (format "id: %s sync insert %s times user %s ms" id run-times (- (System/currentTimeMillis) start)))))

(defn -main
  []
  (let [barrier (CyclicBarrier. threads)
        cdl (CountDownLatch. threads)]
    (dotimes [i threads]
      (.start
       (Thread. (fn []
                  (.await barrier)
                  (sync-insert (* i run-times) cdl)))))
    (.await cdl)
    (drop-all)
    (println "finsh.")))

(defn sync-handler [req]
  (try
    (-> (.getDatabase mongo-client test-db)
        (.getCollection test-class)
        (.insertOne (Document. {"name" (str (rand-int 10000))
                                "_id" (str (System/nanoTime) (rand-int 1000))})))
    {:status 200 :body "success"}
    (catch Exception e
      {:status 400 :body "error"})))


src/benchmark/async.clj
(ns benchmark.async
  (:import  [clojure.lang IPersistentMap Named Keyword Ratio]
            [java.util List Map Date Set]
            [com.mongodb ConnectionString
             ServerAddress
             Block ReadPreference]
            [com.mongodb.async.client MongoClient
             MongoClients MongoClientSettings Observables
             MongoCollection MongoDatabase]
            [com.mongodb.connection ClusterSettings ConnectionPoolSettings]
            [com.mongodb.connection.netty NettyStreamFactoryFactory]
            [com.mongodb.async SingleResultCallback]
            [org.bson Document]
            [org.bson.types ObjectId]
            [java.util.concurrent CyclicBarrier CountDownLatch]))

(def mongo-client (MongoClients/create
                   (-> (MongoClientSettings/builder)
                       (.clusterSettings (-> (ClusterSettings/builder)
                                             (.hosts [(ServerAddress. "192.168.33.10" 27017)])
                                             (.build)))
                       (.connectionPoolSettings (-> (ConnectionPoolSettings/builder)
                                                    (.maxSize 100)
                                                    (.minSize 50)
                                                    (.maxWaitQueueSize 10000)
                                                    (.build)))
                       (.build))))

(def all-run-times 10000)
(def threads 20)
(def run-times (/ all-run-times threads))

(def test-db "test-db")
(def test-class "test-class")

(def default-cb
  (reify
    SingleResultCallback
    (onResult [this result t]
      (if (or result t)
        (println "error?" result t)))))

(defn drop-all
  []
  (let [db-collection (-> (.getDatabase mongo-client test-db)
                          (.getCollection test-class))]
    (.drop db-collection default-cb)))

(defn async-query
  []
  (let [x (promise)
        y (atom 0)
        default-blockcb (reify
                          Block
                          (apply [this s]
                            (when (= (swap! y inc) run-times)
                              (deliver x true))))
        db-client (.getDatabase mongo-client test-db)
        query (Document. {"name" "_User"})
        start (System/currentTimeMillis)]
    (dotimes [_ run-times]
      (-> db-client
          (.getCollection test-class)
          (.find query)
          (.forEach default-blockcb default-cb)))
    @x
    (println  (format "async query %s times use %s ms %s" run-times (- (System/currentTimeMillis) start) @y))))



(defn async-insert
  [id cdl]
  (let [x (promise)
        y (atom 0)
        thread-ids (atom #{})
        cb (reify
             SingleResultCallback
             (onResult [this result t]
               (swap! thread-ids conj (.getId (Thread/currentThread)))
               (when t
                 (println "error:" t))
               (when (= (swap! y inc) run-times)
                 (deliver x true))))
        db-client (.getDatabase mongo-client test-db)
        start (System/currentTimeMillis)]
    (dotimes [i run-times]
      (-> db-client
          (.getCollection test-class)
          (.insertOne (Document. {"name" (str (rand-int 10000))
                                  "_id" (str (+ i id))}) cb)))
    @x
    (.countDown cdl)
    (println  (format "id: %s async insert %s times use %s ms" id run-times (- (System/currentTimeMillis) start)))))

(defn -main
  []
  (let [cdl (CountDownLatch. threads)
        barrier (CyclicBarrier. threads)]
    (dotimes [i threads]
      (.start
       (Thread. (fn []
                  (.await barrier)
                  (async-insert (* i run-times) cdl)))))
    (.await cdl)
    (drop-all)
    (println "finsh.")))

(defn async-handler [req respond raise]
  (let [cb (reify
             SingleResultCallback
             (onResult [this result t]
               (if t
                 (respond {:status 400 :body "error"})
                 (respond {:status 200 :body "success"}))))]
    (-> (.getDatabase mongo-client test-db)
        (.getCollection test-class)
        (.insertOne (Document. {"name" (str (rand-int 10000))
                                "_id" (str (System/nanoTime) (rand-int 1000))}) cb))))
